File: Spelling Bee Helper.R
options(warn = -1)
  library(XML)
  library(curl)
  library(magrittr)
  library(stringr)
  library(tibble)
  library(dplyr)
  library(purrr)
options(warn = 0)

clean.up <- TRUE
  
w <- "Idefgty" %>%
  toupper %>% 
  str_extract_all(., "[A-Z]") %>% 
  unlist %>% 
  unique 

word.url <- "https://norvig.com/ngrams/enable1.txt"
# word.url <- "http://www-personal.umich.edu/~jlawler/wordlist"
# word.url <- "http://www.mieliestronk.com/corncob_caps.txt"

center <- w[1]
others <- setdiff(w, center) %>% sample

stopifnot(length(w) == 7 )
stopifnot(length(setdiff(others, center) ) == 6)

if (!exists("word.list")) {
  word.list.raw <- curl(word.url)
  
  word.list <- word.list.raw %>% 
    readLines(., warn = FALSE) %>% 
    unlist %>%
    toupper %>%
    tibble %>% 
    transmute(., "word" = .) %>% 
    mutate(word.length = str_length(word) ) %>%
    filter(word.length >= 4) %>%
    filter(!grepl("\\W", word) ) %>%
    filter(!str_detect(word, "S")) %>%
    mutate(letters = str_split(word, ""), letters = map(letters, unique) ) %>%
    mutate(unique.letters = lengths(letters) ) %>%
    mutate(points = ifelse(word.length == 4, 1, word.length) + 7 * (unique.letters == 7) ) %>%
    filter(unique.letters <= 7) %>% 
    select(word, letters, word.length, points, unique.letters) %>% 
    arrange(desc(points))
  
    close(word.list.raw)
    rm(word.list.raw)
}

get.words <- function(center.letter, other.letters) {
  require(magrittr)
  require(stringr)
  require(purrr)
  require(dplyr)
  
  all.letters <- c(center.letter, other.letters)
  HC <- word.list %>%
    filter(str_detect(word, center.letter)) %>%
    mutate(invalid.letters = map(letters, setdiff, all.letters)) %>%
    filter(lengths(invalid.letters) == 0) %>%
    select(word, points)
  return(HC)
}


HC.Words <- get.words(center, others)

rm(word.url)

g01 <- others %>% 
  sort %>% 
  paste(., collapse = "") %>% 
  paste0(center, .) %>% 
  tolower

g02 <- c(center, others) %>% 
  sort %>% 
  paste(., collapse = "")


sum(HC.Words$points)
HC.Words %>% 
  filter(points >= 14) %>%
  nrow
# g01
# g02


if (clean.up) {
  rm(w, center, others)
  # rm(word.list)
  rm(g01, g02)
  # rm(HC.words)
}  
  
rm(clean.up)

File: Spelling Bee Helper.py
import pandas as pd
import string as st
import re
import sys
from urllib.request import urlopen


sb_letters = 'iradbol'
# Center Letter Should Always be 1st
letters_reqd = 7
u = 'https://norvig.com/ngrams/enable1.txt'
# u = 'http://www-personal.umich.edu/~jlahcer/wordlist'
# u = 'http://www.mieliestronk.com/corncob_caps.txt'


def unique_letters(word):
    a = set(word)
    b = len(a)
    return b


def spelling_bee_points(wrd):
    ul = unique_letters(wrd)
    wl = len(wrd)
    pts = wl if wl > 4 else 1
    panagram = (ul == 7) * 7
    pts = pts + panagram
    return pts


def build_word_list(url):
    f = urlopen(url)
    w = f.read().decode('utf-8').upper().split()
    w = list(filter(lambda word: len(word) >= 4, w))
    w = list(filter(lambda word: "S" not in (word), w))
    u = list(map(unique_letters, w))
    p = list(map(spelling_bee_points, w))
    d = {'Word': w, 'Unique Letters': u, 'Score': p}
    df = pd.DataFrame(d)
    df = df[(df['Unique Letters']) <= 7]
    df = df.sort_values(['Score', 'Word'], ascending=(False, True))
    return df


def sb_helper(letters, mwl):
    ltrs = letters.upper()
    ltrs = re.findall("[A-Z]", ltrs)
    center = ltrs[0]
    hc = mwl[mwl['Word'].str.contains(center)]
    alphabet = set(st.ascii_uppercase)
    bad_letters = alphabet.difference(set(ltrs))
    for bd_lttr in bad_letters:
        hc = hc[~hc['Word'].str.contains(bd_lttr)]
    cols = [0, 2]
    hc = hc[hc.columns[cols]]
    return hc


reg_ex = re.compile('[^A-Za-z]')
tester = reg_ex.sub('', sb_letters)
bail_out = unique_letters(tester) != letters_reqd
if bail_out:
    sys.exit("Wrong Amount of Letters")
word_list = build_word_list(u)
sb_soln = sb_helper(sb_letters, word_list)


panagrams = sb_soln[sb_soln['Score'] >= 14]
print()
print(f"Panagrams: {len(panagrams)}")
print(f"Total Points: {sum(sb_soln['Score'])}")
print(f"Words: {len(sb_soln)}")
print(sb_soln.to_string())
